app/models/journal.rb

```diff
@@ -56,6 +56,29 @@ class Journal < ActiveRecord::Base
   # logs like the history on issue#show
   scope :changing, -> { where(['version > 1']) }
 
+  # Ensure that no INSERT/UPDATE/DELETE statements as well as other code inside :with_write_lock
+  # is run concurrently to the code inside this block, by using database locking.
+  # Note for PostgreSQL: If this is called from inside a transaction, the lock will last until the
+  #   end of that transaction.
+  # Note for MySQL: THis method does not currently change anything (no locking at all)
+  def self.with_write_lock
+    if OpenProject::Database.mysql?
+      Journal.transaction do
+        # MySQL is very weak when combining transactions and locks. Using an emulation layer to
+        # automatically release an advisory lock at the end of the transaction
+        # FIXME: still creates duplicates... because MySQL defaults to READ REPEATABLE isolation
+        # we need READ COMMITED, which is also the default for PostgreSQL
```

Comment:
The FIXME is no longer up to date right? https://github.com/opf/openproject/pull/3482/files#diff-61fefd1ac4f13f6ec6cebc79a052988dR36